### PR TITLE
Refactor C# discovery snippetgen for MVVM

### PR DESCRIPTION
**Pull request by saicheems**

### CODE REVIEW

File: src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_taskqueue.v1beta2.json.baseline [hunk 1]
```diff
@@ -7,39 +7,31 @@
 // 2. Install the C# client library by adding a dependency on the relevant NuGet
 //    package. Libraries published by Google are owned by google-apis-packages:
 //    https://www.nuget.org/profiles/google-apis-packages
-
 using Google.Apis.Auth.OAuth2;
```
Comment (tcoffee-google) [inline src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_taskqueue.v1beta2.json.baseline]: IIRC, in some languages we have been leaving a blank line after the header comment; we should probably be consistent unless there are known differences in idiom for particular languages.

File: src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_taskqueue.v1beta2.json.baseline [hunk 2]
```diff
@@ -7,39 +7,31 @@
...
 using Google.Apis.Services;
 using Google.Apis.Taskqueue.v1beta2;
 using System;
-using System.Threading.Tasks;
 
 using Data = Google.Apis.Taskqueue.v1beta2.Data;
```
Comment (tcoffee-google) [inline src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_taskqueue.v1beta2.json.baseline]: Is this ordering/spacing for `using Data ...` intentional, or just an artifact of assembly?

File: src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_taskqueue.v1beta2.json.baseline [hunk 3]
```diff
@@ -51,11 +43,12 @@ namespace TaskqueueSample
...
-            // TODO: Change placeholder below to get authentication credentials.
-            // See: https://foo.com/bar
+        public static UserCredential GetCredential()
+        {
+            // TODO: Change placeholder below to generate authentication credentials. See:
+            // https://foo.com/bar
             //
-            // To use the `Get` method, authorize using one of the following scopes:
+            // Authorize using one of the following scopes:
             //   https://www.googleapis.com/auth/taskqueue
```
Comment (tcoffee-google) [inline src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_taskqueue.v1beta2.json.baseline]: Seems best for this indentation width to follow that for the language.

File: src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_taskqueue.v1beta2.json.baseline (hunk 2 above)
Comment (saicheems) [inline src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_taskqueue.v1beta2.json.baseline]: It's intentional - I believe that it's the correct style

File: src/main/java/com/google/api/codegen/discovery/config/ApiaryConfigToSampleConfigConverter.java [hunk 4]
```diff
@@ -104,7 +106,7 @@ private MethodInfo createMethod(Method method) {
     boolean isPageStreaming = isPageStreaming(method);
     FieldInfo pageStreamingResourceField = null;
     if (isPageStreaming) {
-      Type containerType = apiaryConfig.getType(responseTypeUrl);
+      Type containerType = apiaryConfig.getType(method.getResponseTypeUrl());
```
Comment (saicheems) [inline src/main/java/com/google/api/codegen/discovery/config/ApiaryConfigToSampleConfigConverter.java]: It... doesn't actually do anything really, I think this was left over from my work and I forgot to change it. Reverted

File: src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_taskqueue.v1beta2.json.baseline (hunk 1 above)
Comment (saicheems) [inline src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_taskqueue.v1beta2.json.baseline]: I put it back

File: src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_taskqueue.v1beta2.json.baseline (hunk 3 above)
Comment (saicheems) [inline src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_taskqueue.v1beta2.json.baseline]: Changed it to 4 relative spaces

File: src/main/java/com/google/api/codegen/discovery/transformer/csharp/CSharpSampleTypeNameConverter.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+import com.google.api.codegen.util.csharp.CSharpTypeTable;
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableMap;
+import com.google.protobuf.Field;
+
+/** Maps SampleConfig and TypeInfo instances to C# specific TypeName instances. */
+class CSharpSampleTypeNameConverter implements SampleTypeNameConverter {
+
+  /** A map from primitive types in proto to C# counterparts. */
+  private static final ImmutableMap<Field.Kind, String> PRIMIVITVE_TYPE_MAP =
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/discovery/transformer/csharp/CSharpSampleTypeNameConverter.java]: s/PRIMIVITVE/PRIMITVE/ here and below

File: src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_clouduseraccounts.beta.json.baseline [hunk 6]
```diff
@@ -17,44 +19,41 @@ using Google.Apis.CloudUserAccounts.beta;
...
+
                 ApplicationName = "Google-CloudUserAccountsSample/0.1",
             });
 
-
-            // TODO: Change placeholders below to desired parameter values for the `Delete` method:
-
             // Project ID for this request.
-            string project = "{MY-PROJECT}";
+            string project = "";  // TODO: Update placeholder value.
```
Comment (vchudnov-g) [inline src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_clouduseraccounts.beta.json.baseline]: I think I mentioned this in a previous review: there is no placeholder here, so "update placeholder value" seems incorrect. We should either add a placeholder or change the TODO to say "Provide value".

I'm OK fixing this consistently across languages in a separate PR.

File: src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_adexchangebuyer.v1.4.json.baseline [hunk 7]
```diff
@@ -12,31 +12,24 @@ using Google.Apis.AdExchangeBuyer.v1_4;
...
     {
         public static void Main(string[] args)
         {
-            UserCredential credential = GetCredential();
-
             AdExchangeBuyerService adExchangeBuyerService = new AdExchangeBuyerService(new BaseClientService.Initializer
             {
-                HttpClientInitializer = credential,
+                HttpClientInitializer = GetCredential(),
+
```
Comment (tcoffee-google) [inline src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_adexchangebuyer.v1.4.json.baseline]: This blank line seems unnecessary.

File: src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_clouduseraccounts.beta.json.baseline (hunk 6 above)
Comment (tcoffee-google) [inline src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_clouduseraccounts.beta.json.baseline]: My response in said previous review was that the empty string is the placeholder value.

File: src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_translate.v2.json.baseline [hunk 8]
```diff
@@ -26,18 +23,13 @@ namespace TranslateSample
         {
             TranslateService translateService = new TranslateService(new BaseClientService.Initializer
             {
-                ApplicationName = "Google-TranslateSample/0.1",
+                ApiKey = "",  // Update placeholder with desired API key.
```
Comment (tcoffee-google) [inline src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_translate.v2.json.baseline]: This blank line also seems unnecessary.

File: src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_translate.v2.json.baseline [hunk 9]
```diff
@@ -26,18 +23,13 @@ namespace TranslateSample
...
```
Comment (tcoffee-google) [inline src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_translate.v2.json.baseline]: TODO here?

File: src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_translate.v2.json.baseline [hunk 10]
```diff
@@ -26,18 +23,13 @@ namespace TranslateSample
...
+                ApplicationName = "Google-TranslateSample/0.1",
             });
 
-
-            // TODO: Change placeholder below to desired parameter value for the `List` method:
-
             // The text to detect
-            List<string> q = new List<string>();
-
+            List<string> q = new List<string>();  // TODO: Update placeholder value.
```
Comment (tcoffee-google) [inline src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_translate.v2.json.baseline]: Should we use an initializer here (and in Java) that's more readily extensible (to add values)?

File: src/main/resources/com/google/api/codegen/csharp/sample.snip [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+                // To execute asynchronously in an async method, replace `{@class.requestVarName}.Execute()` as shown:
+                {@class.responseVarName} = {@class.requestVarName}.Execute();
+                // {@class.responseVarName} = await {@class.requestVarName}.ExecuteAsync();
+
+                if ({@class.responseVarName}.{@class.pageStreaming.resourceFieldName} == null)
+                {
+                    continue;
+                }
+                @if class.pageStreaming.isResourceMap
+                    foreach ({@class.pageStreaming.resourceElementTypeName} {@class.pageStreaming.resourceVarName} in response.{@class.pageStreaming.resourceFieldName}) {
```
Comment (tcoffee-google) [inline src/main/resources/com/google/api/codegen/csharp/sample.snip]: Put open brace on next line.

**Pull request merged**